scipy/linalg/tests/test_lapack.py

```diff
@@ -1793,6 +1793,188 @@ def test_getc2_gesc2():
                                       x/scale, decimal=4)
 
 
+@pytest.mark.parametrize('size', [(6, 5), (5, 5)])
+@pytest.mark.parametrize('dtype', DTYPES)
+@pytest.mark.parametrize('joba', ['C', 'E', 'F', 'G', 'A', 'R'])
+@pytest.mark.parametrize('jobu', ['U', 'F', 'W', 'N'])
+@pytest.mark.parametrize('jobv', ['V', 'J', 'W', 'N'])
+@pytest.mark.parametrize('jobr', ['R', 'N'])
+@pytest.mark.parametrize('jobt', ['T', 'N'])
+@pytest.mark.parametrize('jobp', ['P', 'N'])
+def test_gejsv_general(size, dtype, joba, jobu, jobv, jobr, jobt, jobp):
+    """Test the lapack routine ?gejsv.
+
+    This function tests that a singular value decomposition can be performed
+    on the random M-by-N matrix A. The test performs the SVD using ?gejsv
+    then performs the following checks:
+
+    * ?gejsv exist successfully (info == 0)
+    * The returned singular values are correct
+    * `A` can be reconstructed from `u`, `SIGMA`, `v`
+    * Ensure that u.T @ u is the identity matrix
+    * Ensure that v.T @ v is the identity matrix
+    * The reported matrix rank
+    * The reported number of singular values
+    * If denormalized floats are required
+
+    Notes
+    -----
+    joba specifies several choices effecting the calculation's accuracy
+    Although all arguments are tested, the tests only check that the correct
+    solution is returned - NOT that the prescribed actions are performed
+    internally.
+    """
+    seed(42)
+
+    # Define some constants for later use:
+    m, n = size
+    atol = 100 * np.finfo(dtype).eps
+    A = generate_random_dtype_array(size, dtype)
+    gejsv = get_lapack_funcs('gejsv', dtype=dtype)
+
+    # Set up checks for invalid job? combinations
+    # if an invalid combination occurs we set the appropriate
+    # exit status.
+    lsvec = jobu in {'U', 'F'}  # Calculate left singular vectors
+    rsvec = jobv in {'V', 'J'}  # Calculate right singular vectors
+    l2tran = (jobt == 'T') and (m == n)
+    is_complex = np.iscomplexobj(A)
+
+    invalid_real_jobv = (jobv == 'J') and (not lsvec) and (not is_complex)
+    invalid_cplx_jobu = (jobu == 'W') and not (rsvec and l2tran) and is_complex
+    invalid_cplx_jobv = (jobv == 'W') and not (lsvec and l2tran) and is_complex
```

Comment:
These checks may be overly complicated. I've tried to duplicate the logic as closely as possible as from the [reference implementation](http://www.netlib.org/lapack/explore-html/d7/d36/cgejsv_8f_source.html#l00634)